aten/src/THS/generic/THSTensor.cpp

```diff
@@ -135,128 +135,138 @@ THSTensor *THSTensor_(new)(void)
 /* Pointer-copy init */
 THSTensor *THSTensor_(newWithTensor)(THLongTensor *indices, THTensor *values)
 {
-  return THSTensor_(newWithTensorAndSize)(indices, values, NULL);
+  // If sizes are not given, it is inferred as max index of each dim.
+  int64_t nDimI = THLongTensor_size(indices, 0);
+  int64_t nDimV = THTensor_(nDimension)(values) - 1;
+
+  THSTensor *self = THSTensor_(new)();
+
+  THLongTensor *ignore = THLongTensor_new();
+  THLongTensor *computed_indices_sizes = THLongTensor_new();
+  THLongTensor *computed_sizes = THLongTensor_newWithSize1d(nDimI + nDimV);
+  THLongTensor_max(computed_indices_sizes, ignore, indices, 1, 1);
+  THLongTensor_add(computed_indices_sizes, computed_indices_sizes, 1);
+  for (int d = 0; d < nDimI; d++) {
+      THTensor_fastSet1d(computed_sizes, d, THTensor_fastGet1d(computed_indices_sizes, d));
+  }
+  for (int d = 0; d < nDimV; d++) {
+      THTensor_fastSet1d(computed_sizes, nDimI + d, THTensor_(size)(values, d + 1));
+  }
+  THSTensor_(rawResize)(self, nDimI, nDimV, THLongTensor_data(computed_sizes));
+  THLongTensor_free(computed_indices_sizes);
+  THLongTensor_free(computed_sizes);
+  THLongTensor_free(ignore);
+
+  // NB: by default, we do NOT clone indices/values into the sparse tensor.
+  // Efficient API by default!
+  THSTensor_(_move)(self, THLongTensor_newWithTensor(indices), THTensor_(newWithTensor)(values));
```

Comment:
Can we not have this three times? Is it possible to call `newWithTensorAndSizeUnsafe` in this can `newWithTensorAndSize`?

Reply:
It's possible, but it would look something like the old newWithTensorSize function (because the checking happens in the middle of the function, it's hard to extract). I thought that the repetition of this one line was a better choice than having the confusing code paths.

Reply:
It is actually three lines
```
THSTensor *self = THSTensor_(new)();
THSTensor_(rawResize)(self, nDimI, nDimV, THLongTensor_data(computed_sizes));
THSTensor_(_move)(self, THLongTensor_newWithTensor(indices), THTensor_(newWithTensor)(values));
```
With a `NB`, I would prefer to not replicate this...

Reply:
How about refactoring those three lines into another function?

Since the checks happen after `if (sizes == NULL)` and `if (THLongTensor_nDimension(indices) == 0 && THTensor_(nDimension)(values) == 0)`, and those are required in both the safe and unsafe versions, it doesn't feel good to call the unsafe version from the safe one.

Reply:
A local helper function would be great :)